### DISCUSSION TITLE
v0.23.0-rc7: First sign in with Oauth does not set verified to true, second one does

### DISCUSSION THREAD

**Discussion by Plorenzo** [pocketbase/pocketbase#5703] [Q&A]
When a new user signs up with Oauth2 either of this:

```go
const authData = await pb.collection('users').authWithOAuth2({ provider: 'google' });
const authData = await pb.collection('users').authWithOAuth2({ provider: 'apple' });
```

The user gets created but it won't get logged in if the requirement for email verified is set to true and I get this error instead.  

```
ERROR POST /api/collections/users/auth-with-oauth2
└─ The request doesn't satisfy the collection requirements to authenticate.
```
If I click the sign in button again, it will mark the email as verified this time and the user will log in just fine. 

I have 2 hooks, this one before creating the user
```go
app.OnRecordCreateRequest("users").BindFunc(func(e *core.RecordRequestEvent) error {

		e.Record.Set("role", "user")
		e.Record.Set("status", "active")
		return e.Next()
	})
```
which produces this INSERT 
```
[0.00ms] INSERT INTO `users` (`avatar`, `bio`, `created`, `email`, `emailVisibility`, `experience`, `id`, `name`, `password`, `role`, `status`, `tokenKey`, `updated`, `username`, `verified`) VALUES ('acg8oc_k_ps0qqe_ktfm81_gto_ur_hq4r3_tv_mfx_iovm_idde_m7_yv0_lgd1_x_lm_q_s96_c_sujxs0vp60_r5nszbss6s.jpg', '', '', 'REDACTED@gmail.com', false, '', '_temp_wWVuyiJMC01ZSqe', 'Pablo Lorenzo', '', 'user', 'active', '', '', '', false)
```

and a second one 
```go
app.OnRecordAfterCreateSuccess("users").BindFunc(func(e *core.RecordEvent) error {
		// Adds user to usersStatus collection. This is needed because accountStatus should be only edited by admins
		// and if it lives as a fiels in user we can't control that permission.

		app.Logger().Debug("OnRecordAfterCreateSuccess Users")
		record := e.Record
		record.Set("avatar", embeddedFile)

		err := app.Save(record)
		if err != nil {
			return err
		}
               return e.Next()
	})
```

which produces another INSERT 
```
[0.00ms] INSERT INTO `users` (`avatar`, `bio`, `created`, `email`, `emailVisibility`, `experience`, `id`, `name`, `password`, `role`, `status`, `tokenKey`, `updated`, `username`, `verified`) VALUES ('acg8oc_k_ps0qqe_ktfm81_gto_ur_hq4r3_tv_mfx_iovm_idde_m7_yv0_lgd1_x_lm_q_s96_c_sujxs0vp60_r5nszbss6s.jpg', '', '2024-10-19 10:19:07.673Z', 'redacted@gmail.com', false, '', '46205cx1tu324s6', 'Pablo Lorenzo', 'redacted', 'user', 'active', 'redacted', '2024-10-19 10:19:07.673Z', 'kxsb50fw1z08rh57ki1594239g6266', false)
```

Not sure if this the expected behaviour and I would need to set verified to true on first log in manually or if there is something wrong .

**Comment by pnmcosta** [comment]
pretty sure it's the additional `app.Save(record)` on OnRecordAfterCreateSuccess func that is duplicating the insert. You shouldn't need to manually insert there. Please note that:

> Similar to the new hook handlers, middlewares must call e.Next() in order to proceed with the execution chain.

From: https://pocketbase.io/v023upgrade/go/#event-hooks-changes

**Comment by ganigeorgiev** [comment]
The `verified` state is set automatically but there are several requirements depending on whether the PocketBase user is new or the OAuth2 is just link to an existing one (for example you are already authenticated before the authWithOAuth2 call), whether the provider returns a verified email field, etc.

With that said, I think your issue comes from the custom hooks and specifically the second save in `OnRecordAfterCreateSuccess` because `e.Record` could contain an old state from before the OAuth2 verified update and with the second save you are enforcing the data from that time (the OAuth2 user creation executes in a transaction and the **After** success hooks are **delayed** until the transaction completes).

Unfortunately your code example seems incomplete and I don't understand what is the desired behavior to recommend a proper workaround.

---

The underlying issue is a little more complicated to explain and I'm not sure how to do it properly as I haven't started documenting that part yet.

Technically with the new hooks design you can have an after execution using simply `OnRecordCreate`/`OnRecordCreateRequest` and add your after calls _after_ the `e.Next()` invocation:

```go
app.OnRecordCreate("users").BindFunc(func(e *core.RecordEvent) error {
    if err := e.Next(); err != nil {
            return err 
    }

    // executes right after the app.Save() call ....

    return nil
})
```

As probably expected, everything after `e.Next()` in the above will happen after `app.Save()` BUT because the save operation could be part of a transaction there is no guarantee that the changes are actually persisted and that later the same transaction is not rollbacked due to an error or some other external factor. For example you can have a situation like:

```go
app.RunInTransaction(func(txApp core.App) error {
    record, _ := txApp.FindRecordById("users", "RECORD_ID")
    record.Set("name", "test")
    err := txApp.Save(record)
    if err != nil {
        return err
    }

    // ....

    return errors.New("rollback transaction")
})
```

And that is the reason why there are special DB **After** success/error hooks.
They are triggered after the specific operation (create/update/delete) but their execution is **delayed** until the record change is persisted (or not in case of error). For the after success this mean until the parent transaction has been completed (if there is no transaction it executes directly right away).

The problem is that the `e.Record` in the **After** success hooks points to a record at the moment when they were triggered, which could cause inconsistencies if there is another save operation of the same record after the initial one that uses a different record pointer.

I'll have to investigate it in more details before the final release and eventually will see if we can "autorefresh" the record before executing the after hook but for now my recommendation is to try to avoid the double save (_or if you can't refetch the record from the database on your own_).

**Reply by pnmcosta** [reply]
> The problem is that the e.Record in the After success hooks points to a record at the moment when they were triggered

This is interesting and good to know. 

I think, and I have a similar use case, that if the OP wants the record to include fields before the record is actually saved would be better to bind to `OnRecordCreateExecute`, correct? This occurs after validation, which you would need to be mindful of.

I would probably do:

```
app.OnRecordCreateExecute("users").BindFunc(func(e *core.RecordEvent) error {
  e.Record.Set("role", "user")
  e.Record.Set("status", "active")
  e.Record.Set("avatar", embeddedFile)

  return e.Next()
})
```

To ensure the fields are committed to DB.

**Comment by Plorenzo** [comment]
Thanks both for the answers. Indeed removing `err := app.Save(record)` from `app.OnRecordAfterCreateSuccess` fixes the issue, the second INSERT was overwriting the record 

>I don't understand what is the desired behavior 

When a new user is created I need to set two mandatory fields before the user is saved `status` and `role`, I can't use the `OnRecordAfterCreateSuccess` because it will fail saying that are missing required fields so that's why Im using `OnRecordCreateRequest`

>I would probably do:

Im using `OnRecordAfterCreateSuccess` for the avatar because is a less critical task and even if something goes wrong the user will be saved.

**Reply by pnmcosta** [reply]
In that case, having to use `OnRecordAfterCreateSuccess`, then you should run your save, but after all other `OnRecordAfterCreateSuccess` have executed:

```
app.OnRecordAfterCreateSuccess("users").BindFunc(func(e *core.RecordEvent) error {
    if err := e.Next(); err != nil {
            return err 
    }
    
    // EDIT: note from @ganigeorgiev comment: https://github.com/pocketbase/pocketbase/discussions/5703#discussioncomment-10990978
    // > The problem is that the e.Record in the After success hooks points to a record at the moment when they were triggered
  
    // you may want to app.FindRecordById here to have a clean copy of the record from DB instead of using e.Record

    // executes right after the original app.Save() call (INSERT record)
    e.Record.Set("avatar", embeddedFile)
    
    // UPDATE record
    return  app.Save(record)
})
```
